### DISCUSSION TITLE
How to upsert when using `EntityFactory`?

### DISCUSSION THREAD

**Discussion by squivix** [mikro-orm/mikro-orm#3968] [Q&A]
I am using `EntityFactory` with random data for my automated testing, but I keep getting unique constraint violations for certain fields. I don't want my tests to sometimes fail because the database they started with was not clean, and truncating the datbase before every test adds a lot of overhead. It would be very nice if I could ignore (or better upsert on) unique constraint violations when testing.

I tried overriding `createOne` and `create` to use `em.upsert` instead of `em.create` . Here's the code for that:

```ts
export abstract class CustomFactory<T extends Object> extends Factory<T> {
    shouldUpsert: boolean

    constructor(em: EntityManager, shouldUpsert?: boolean) {
        super(em);
        this.shouldUpsert = shouldUpsert ?? false;
    }
    async createOne(overrideParameters?: EntityData<T>): Promise<T> {
        const entity = ((this as any).makeEntity(overrideParameters)) as T;
        if (this.shouldUpsert)
            await ((this as any).em as EntityManager).upsert(this.model, entity);
        else {
            (this as any).em.persist(entity);
            await (this as any).em.flush();
        }
        return entity;
    }

    async create(amount: number, overrideParameters?: EntityData<T>): Promise<T[]> {
        const entities = [...Array(amount)].map(() => {
            return (this as any).makeEntity(overrideParameters) as T;
        });
        if (this.shouldUpsert) {
            await ((this as any).em as EntityManager).upsertMany(this.model, entities);
        } else {
            (this as any).em.persist(entities);
            await (this as any).em.flush();
        }
        return entities;
    }
}
```

I keep getting `Error: Unique property value required for upsert, provide one of: id` which seems to be triggered by an entity that doesn't have any unique fields besides the id, and the value of the id is not provided with factories. Thing is though, I have many `EntityFactories` that rely on each other, so I sometimes want to call `upsert` and sometimes not. Maybe I would need a way to know if the entity has unique fields besides the id to include it.

Please let me know if there's a better way of doing this. I might be writing more code than neccessary.

**Comment by B4nan** [comment]
What about ensuring your unique fields are really unique? E.g. adding random suffix or a timestamp to the values.

Otherwise, I would say you should wipe the database before you run your seeders, sounds like you want to run them multiple times without wiping which I would personally discourage. You don't need to do it on each test, that is up to you, but doing those two actions out of sync is weird.

For the problem you have, you could check entity metadata dynamically and use upsert only when there are some unique properties other than a autoincrement PK. 

```ts
const meta = em.getMetadata().get(this.model);
const canUseUpsert = meta.uniqueProps.length > 0;
```

**Reply by squivix** [reply]
I do wipe the database before running all tests, but not before running each test. My unique fields are very short (2 characters), and I have many tests, so the chance of collision is high even within one run.

Using this metadata check I am able to avoid collisions when creating an instance of the entity with the problematic field, but not when creating other entities which depend on it. For example if I have:

```ts

export class EntityAFactory extends CustomFactory<EntityA> {
    readonly model = EntityA;

    protected definition(faker: Faker): EntityData<EntityA> {
      return {
          collsionField: faker.random.alpha({count: 2}),
      };
    }
}
```

and 


```ts

export class EntityBFactory extends CustomFactory<EntityB> {
    readonly model = EntityB;

    protected definition(faker: Faker): EntityData<EntityB> {
      return {
          entityA: {
              collsionField: faker.random.alpha({count: 2}),
          }
      };
    }
}
```

Then when I do:

```ts
const entityAs = await entityAFactory.create(10);
```
I do avoid collisions, but when I do:

```ts
const entityBs = await entityBFactory.create(10);
```

then I don't avoid collisions on Entity A, because the EntityB does not have any `uniqueProps`. Is there a way to check if any sub-entity has unique props?

Other than that, what about ignoring collisions? Do I also need to have uniqueProps in order to do that?

**Reply by B4nan** [reply]
> , but not before running each test. 

But why would you seed before each test then? Those two operations should be tied together, it does not make sense to wipe once and seed multiple times.

> Is there a way to check if any sub-entity has unique props?

You can check it the same way via metadata, you have `meta.relations` as array of relation properties, and those have `prop.targetMeta` with the metadata of entity they target.
